File: src/Offering/Offering/Models/Discount.cs
namespace Offering.Models;

public class Discount
{
    public decimal DiscountRate { get; private set; }
    
    public DateTime StartDate { get; private set; }
    
    public DateTime EndDate { get; private set; }
    
    protected Discount(decimal discountRate, DateTime startDate, DateTime endDate)
    {
        DiscountRate = discountRate;
        StartDate = startDate;
        EndDate = endDate;
    }
    
    public static Discount Create(decimal discountRate, DateTime startDate, DateTime endDate)
    {
        return new Discount(discountRate, startDate, endDate);
    }
}

File: src/OfferSB/src/main/java/com/example/offersb/models/Discount.java
package com.example.offersb.models;

import jakarta.persistence.Column;
import jakarta.persistence.Embeddable;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@Embeddable
@NoArgsConstructor
public class Discount {

    @Getter
    private double discountRate;

    @Getter
    @Column(name = "discount_start_date")
    private LocalDateTime startDate;

    @Getter
    @Column(name = "discount_end_date")
    private LocalDateTime endDate;

    private Discount(double discountRate, LocalDateTime startDate, LocalDateTime endDate) {
        this.discountRate = discountRate;
        this.startDate = startDate;
        this.endDate = endDate;
    }

    public static Discount create(double discountRate, LocalDateTime startDate, LocalDateTime endDate) {
        return new Discount(discountRate, startDate, endDate);
    }
}
